src/snowflake/snowpark/mock/_functions.py

```diff
@@ -863,59 +859,142 @@ def mock_to_char(
+            return try_convert(lambda x: converter(x, format), try_cast, data)
+        elif isinstance(source_datatype, TimestampType):
+            default_format = _DEFAULT_OUTPUT_FORMAT.get(TimestampType)
+            (
+                format,
+                fractional_seconds,
+            ) = convert_snowflake_datetime_format(_fmt, default_format)
+            # handle 3f, can use str index
+            time_str = try_convert(
+                lambda x: datetime.date.strftime(x, format), try_cast, data
+            ).strip()
+            # python doesn't offer a way to control digits in microseconds
+            # when converting datatime into string using format %f, hence here we manually control the output fractional
+            # microsecond parts
+            # we find the beginning of 6 consecutive digits and manipulate the string
+            # CAVEAT: this solution can not handle cases yyyymmddhhmmss
+            if "%f" in format:
+                pattern = r"\d{6}"
+                if len(re.findall(pattern, time_str)) > 1:
+                    # if we detect multiple consecutive 6 digits, it's a format we can't handle
+                    LocalTestOOBTelemetryService.get_instance().log_not_supported_error(
+                        external_feature_name=f"Use format string {_fmt} with TimestampType in TO_CHAR",
+                        internal_feature_name="mock_to_char",
+                        parameters_info={
+                            "source_datatype": type(source_datatype).__name__,
+                            "fmt": str(_fmt),
+                        },
+                        raise_error=NotImplementedError,
+                    )
+                start_idx = re.search(r"\d{6}", time_str).start()
+                end_idx = start_idx + 6
+                # truncate the microsecond string
+                fractional_seconds_str = time_str[start_idx:end_idx][
+                    0 : 6 if fractional_seconds >= 6 else fractional_seconds - 6
+                ]
+                # concatenate the whole string
+                time_str = (
+                    time_str[:start_idx]
+                    + fractional_seconds_str
+                    + (time_str[end_idx:] if end_idx < len(time_str) else "")
+                )
+            return time_str
+        elif isinstance(source_datatype, BinaryType):
+            _fmt = (_fmt or "HEX").upper()
+            fmt_decoder = {
+                "HEX": binascii.hexlify,
+                "BASE64": base64.b64encode,
+                "UTF-8": lambda x: x,
+            }.get(_fmt)
+
+            if fmt_decoder is None:
+                raise SnowparkSQLException(f"Invalid binary format {fmt}")
+            return try_convert(fmt_decoder, try_cast, data).decode()
+        elif isinstance(source_datatype, BooleanType):
+            return try_convert(lambda x: str(x).lower(), try_cast, data)
+        elif isinstance(source_datatype, (VariantType, ArrayType, MapType)):
```

Comment:
I tries the following live test with StructType:
```python
    session.create_dataframe([1]).select(
        to_char(
            object_construct(lit("A"), lit("foo"), lit("B"), lit(0.05)).cast(
                StructType([StructField("A", StringType()), StructField("B", DoubleType())],structured=True)
            )
        )
    ).collect()
```

which gives me error `SQL compilation error: E       invalid type [TO_CHAR(STRUCTURED_CAST(OBJECT_CONSTRUCT('A', 'foo', 'B', 0.05)))] for parameter 'TO_CHAR'`.

does it mean to_char doesn't support StructType yet or it's because of my improper dataframe creation?

Reply:
I did some testing. It looks like to_char doesn't support any of the structured types yet, only their semi-structured forms.